File: sdk/python/rcf_core/pdg.py
# NOTICE: This file is protected under RCF-PL
# [RCF:PROTECTED]
"""
PDG — the Program Dependence Graph object from RCF-CORRELATION.md §1.2.

    G = (V, E, ℓ)

  V : operations / values, each labeled ℓ(v) = (cls, op) over Σ.
  E : typed dependence edges (DATA, CTRL; ORDER optional).

Hard invariants (RCF-SIGMA.md §2):
  - identifiers are NOT stored (no names),
  - concrete types are NOT stored,
  - every node label and every edge type is drawn ONLY from Σ.

Construction goes through a Sigma instance so an out-of-alphabet label fails loudly
at build time, not silently at compare time.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from .sigma import Sigma


@dataclass(frozen=True)
class Node:
    """A PDG node. `op` may be None (class-only label is valid per Σ)."""
    nid: int
    cls: str
    op: str | None = None

    @property
    def label(self) -> str:
        """Canonical label string: 'CLASS' or 'CLASS.OP'."""
        return self.cls if self.op is None else f"{self.cls}.{self.op}"


@dataclass(frozen=True)
class Edge:
    src: int
    dst: int
    etype: str  # DATA | CTRL | ORDER


class PDG:
    """A labeled directed dependence graph, validated against Σ on every add."""

    def __init__(self, sigma: Sigma):
        self.sigma = sigma
        self._nodes: dict[int, Node] = {}
        self._edges: list[Edge] = []
        self._next_id = 0

    # ─── construction (Σ-checked) ─────────────────────────────────────────────

    def add_node(self, cls: str, op: str | None = None) -> int:
        """Add a node; raises SigmaError if (cls, op) is not in Σ. Returns its id."""
        self.sigma.require_label(cls, op)
        nid = self._next_id
        self._next_id += 1
        self._nodes[nid] = Node(nid, cls, op)
        return nid

    def add_edge(self, src: int, dst: int, etype: str = "DATA") -> None:
        """Add a typed edge; raises on unknown endpoints or non-Σ edge type."""
        if not self.sigma.is_edge(etype):
            from .sigma import SigmaError
            raise SigmaError(f"edge type not in Σ {self.sigma.version}: {etype!r}")
        if src not in self._nodes or dst not in self._nodes:
            raise KeyError(f"edge endpoint missing: {src}->{dst}")
        self._edges.append(Edge(src, dst, etype))

    # ─── access ───────────────────────────────────────────────────────────────

    @property
    def nodes(self) -> list[Node]:
        return list(self._nodes.values())

    @property
    def edges(self) -> list[Edge]:
        return list(self._edges)

    def node(self, nid: int) -> Node:
        return self._nodes[nid]

    def __len__(self) -> int:
        return len(self._nodes)

    # ─── adjacency (for WL aggregation) ───────────────────────────────────────

    def neighbors(self, nid: int) -> list[tuple[str, str, int]]:
        """
        Typed, directed neighborhood of `nid` for WL aggregation.
        Returns (direction, edge_type, other_id) where direction is 'OUT' for an
        edge nid->other and 'IN' for other->nid. Both directions are kept because
        data/control FLOW has a direction that carries meaning.
        """
        out: list[tuple[str, str, int]] = []
        for e in self._edges:
            if e.src == nid:
                out.append(("OUT", e.etype, e.dst))
            elif e.dst == nid:
                out.append(("IN", e.etype, e.src))
        return out

    # ─── serialization (debug / fingerprint input) ────────────────────────────

    def to_dict(self) -> dict:
        """
        Deterministic dict form. Carries sigma_version AND alphabet_hash so any
        downstream fingerprint is bound to the exact alphabet it was built under
        (RCF-SIGMA.md §7 — alphabet_hash is the comparability key).
        """
        return {
            "sigma_version": self.sigma.version,
            "alphabet_hash": self.sigma.alphabet_hash,
            "nodes": [
                {"id": n.nid, "label": n.label} for n in self._nodes.values()
            ],
            "edges": [
                {"src": e.src, "dst": e.dst, "type": e.etype} for e in self._edges
            ],
        }

    def label_multiset(self) -> dict[str, int]:
        """Frequency of each node label — the WL iteration-0 feature vector φ₀."""
        counts: dict[str, int] = {}
        for n in self._nodes.values():
            counts[n.label] = counts.get(n.label, 0) + 1
        return counts

File: sdk/typescript/src/core/pdg.ts
// NOTICE: This file is protected under RCF-PL
// [RCF:PROTECTED]

import { Sigma, SigmaError } from './sigma.js';

export class PdgNode {
  constructor(
    public readonly nid: number,
    public readonly cls: string,
    public readonly op: string | null = null
  ) {}

  public get label(): string {
    return this.op === null ? this.cls : `${this.cls}.${this.op}`;
  }
}

export class PdgEdge {
  constructor(
    public readonly src: number,
    public readonly dst: number,
    public readonly etype: string // DATA | CTRL | ORDER
  ) {}
}

export class PDG {
  private readonly _nodes = new Map<number, PdgNode>();
  private readonly _edges: PdgEdge[] = [];
  private _nextId = 0;

  constructor(public readonly sigma: Sigma) {}

  public addNode(cls: string, op: string | null = null): number {
    this.sigma.requireLabel(cls, op || undefined);
    const nid = this._nextId++;
    this._nodes.set(nid, new PdgNode(nid, cls, op));
    return nid;
  }

  public addEdge(src: number, dst: number, etype = 'DATA'): void {
    if (!this.sigma.isEdge(etype)) {
      throw new SigmaError(`edge type not in Σ ${this.sigma.version}: ${etype}`);
    }
    if (!this._nodes.has(src) || !this._nodes.has(dst)) {
      throw new Error(`edge endpoint missing: ${src}->${dst}`);
    }
    this._edges.push(new PdgEdge(src, dst, etype));
  }

  public get nodes(): PdgNode[] {
    return Array.from(this._nodes.values());
  }

  public get edges(): PdgEdge[] {
    return [...this._edges];
  }

  public node(nid: number): PdgNode {
    const n = this._nodes.get(nid);
    if (!n) {
      throw new Error(`Node not found: ${nid}`);
    }
    return n;
  }

  public get length(): number {
    return this._nodes.size;
  }

  /**
   * Typed, directed neighborhood of `nid` for WL aggregation.
   * Returns tuple list of [direction, edge_type, other_id].
   */
  public neighbors(nid: number): [string, string, number][] {
    const out: [string, string, number][] = [];
    for (const e of this._edges) {
      if (e.src === nid) {
        out.push(['OUT', e.etype, e.dst]);
      } else if (e.dst === nid) {
        out.push(['IN', e.etype, e.src]);
      }
    }
    return out;
  }

  public toDict(): Record<string, any> {
    return {
      sigma_version: this.sigma.version,
      alphabet_hash: this.sigma.alphabetHash,
      nodes: this.nodes.map((n) => ({ id: n.nid, label: n.label })),
      edges: this._edges.map((e) => ({ src: e.src, dst: e.dst, type: e.etype }))
    };
  }

  public labelMultiset(): Record<string, number> {
    const counts: Record<string, number> = {};
    for (const n of this._nodes.values()) {
      counts[n.label] = (counts[n.label] || 0) + 1;
    }
    return counts;
  }
}
